backend/dataall/modules/dataset_sharing/services/share_managers/s3_bucket_share_manager.py

```diff
@@ -0,0 +1,445 @@
+        )
+        self.source_account_id = target_bucket.AwsAccountId
+        self.target_account_id = target_environment.AwsAccountId
+        self.source_env_admin = source_env_group.environmentIAMRoleArn
+        self.target_requester_IAMRoleName = share.principalIAMRoleName
+        self.bucket_name = target_bucket.S3BucketName
+        self.dataset_admin = dataset.IAMDatasetAdminRoleArn
+        self.bucket_region = target_bucket.region
+
+    @abc.abstractmethod
+    def process_approved_shares(self, *kwargs) -> bool:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def process_revoked_shares(self, *kwargs) -> bool:
+        raise NotImplementedError
+
+    def grant_s3_iam_access(self):
+        """
+        Updates requester IAM role policy to include requested S3 bucket and kms key
+        :return:
+        """
+        logger.info(
+            f'Grant target role {self.target_requester_IAMRoleName} access policy'
+        )
+        existing_policy = IAM.get_role_policy(
+            self.target_account_id,
+            self.target_requester_IAMRoleName,
+            IAM_S3BUCKET_ROLE_POLICY,
+        )
+        key_alias = f"alias/{self.target_bucket.KmsAlias}"
+        kms_client = KmsClient(self.source_account_id, self.source_environment.region)
+        kms_key_id = kms_client.get_key_id(key_alias)
+
+        if existing_policy:  # type dict
+            s3_target_resources = [
+                f"arn:aws:s3:::{self.bucket_name}",
+                f"arn:aws:s3:::{self.bucket_name}/*"
+            ]
+
+            share_manager = ShareManagerUtils(
+                self.session,
+                self.dataset,
+                self.share,
+                self.source_environment,
+                self.target_environment,
+                self.source_env_group,
+                self.env_group
+            )
+            share_manager.add_missing_resources_to_policy_statement(
+                resource_type=self.bucket_name,
+                target_resources=s3_target_resources,
+                existing_policy_statement=existing_policy["Statement"][0],
+                iam_role_policy_name=IAM_S3BUCKET_ROLE_POLICY
+            )
+
+            if kms_key_id:
+                kms_target_resources = [
+                    f"arn:aws:kms:{self.bucket_region}:{self.source_account_id}:key/{kms_key_id}"
+                ]
+
+                share_manager.add_missing_resources_to_policy_statement(
+                    resource_type=kms_key_id,
+                    target_resources=kms_target_resources,
+                    existing_policy_statement=existing_policy["Statement"][1],
+                    iam_role_policy_name=IAM_S3BUCKET_ROLE_POLICY
+                )
+
+            policy = existing_policy
+        else:
+            logger.info(
+                f'{IAM_S3BUCKET_ROLE_POLICY} does not exists for IAM role {self.target_requester_IAMRoleName}, creating...'
+            )
+            policy = {
```

Comment:
The limit for managed IAM policies is 6144 characters and white spaces don't count. I counted the chars of the policy without replacing the bucket_name... and other parameters and we have a policy of ~370 characters. With each new bucket we will add around ~170 chars. Which means that each team can request access to around 35 buckets (very quick and dirty aproximation). Do you think that you will face this scenario?

In case of failure, the block is safely placed inside a try/except block in the process_approved_shares. So in the worst case scenario, the share will end up in the failure state

Reply:
I think we can very easily reach this limit specially for datasets which are to be "open" or used across teams. We should find a workaround for this. Maybe we can have multiple policies ending with policy-1, policy-2 etc as we do for dataallpivotrole right now..

Reply:
I agree. Lets file a separate github issue for this and address it there.